python/dlr/tf_model.py

```diff
@@ -0,0 +1,153 @@
+                raise ValueError("Invalid device type {}. Valid devices: {}".format(dev_type, devices))
+            dev_id = 0 if dev_id is None else dev_id
+            device = "/{}:{}".format(dev_type, dev_id)
+
+        self.graph = self._load_frozen_graph(frozen_graph_file, device)
+        self.input_tensor_names, self.output_tensor_names = self._get_input_and_output_names(self.graph)
+        self.input_values = {}
+
+    @staticmethod
+    def _load_frozen_graph(frozen_graph_file, device):
+        print("Loading frozen graph: {}".format(frozen_graph_file))
+        with tf.device(device):
+            with tf.gfile.GFile(frozen_graph_file, 'rb') as f:
+                graph_def = tf.GraphDef()
+                graph_def.ParseFromString(f.read())
+            with tf.Graph().as_default() as graph:
+                tf.import_graph_def(graph_def, name=PREFIX)
+            return graph
+
+    @staticmethod
+    def _get_input_and_output_names(graph):
+        input_tensor_names = []
+        output_tensor_names = set()
+        op_prefix = PREFIX + "/"
+        for op in graph.get_operations():
+            if not op.name.startswith(op_prefix):
+                continue
+            if op.type == 'Placeholder' and op.inputs.__len__() == 0 and op.outputs.__len__() == 1:
+                input_tensor_names.append(op.outputs[0].name)
+            if op.outputs.__len__() == 1:
+                output_tensor_names.add(op.outputs[0].name)
+        for op in graph.get_operations():
+            for in_t in op.inputs:
+                if in_t.name in output_tensor_names:
+                    output_tensor_names.remove(in_t.name)
+        # Sort list of output tensor names in order to get consistent output in run()
+        output_tensor_names = list(output_tensor_names)
+        output_tensor_names.sort()
+        return input_tensor_names, output_tensor_names
+
+    def _validate_input_name(self, name):
+        if name not in self.input_tensor_names:
+            raise ValueError(
+                "Invalid input tensor name '{}'. List of input tensor names: {}".format(name, self.input_tensor_names))
+
+    def _validate_input(self, input_values):
+        if isinstance(input_values, dict):
+            for k in input_values.keys():
+                if not isinstance(k, str):
+                    raise ValueError("input key must be string")
+                self._validate_input_name(k)
+        else:
+            raise ValueError("input_values must be of type dict")
+
+    def get_input_names(self):
+        """
+        Get all input names
+
+        Returns
+        -------
+        out : list of :py:class:`str`
+        """
+        return self.input_tensor_names
+
+    def get_output_names(self):
+        """
+        Get all output names
+
+        Returns
+        -------
+        out : list of :py:class:`str`
+        """
+        return self.output_tensor_names
+
+    def get_input(self, name, shape=None):
```

Comment:
confused about the function name, maybe get_intput_values?

Reply:
`TFModelImpl` must implement all public method from DLRModel. DLRModel is out public  interface.